lib/go-atscfg/remapdotconfig.go

```diff
@@ -0,0 +1,413 @@
+	} else {
+		text = GetServerConfigRemapDotConfigForEdge(cacheURLConfigParams, dsProfilesCacheKeyConfigParams, serverPackageParamData, serverInfo, remapDSData, atsMajorVersion, hdr)
+	}
+	return text
+}
+
+func GetServerConfigRemapDotConfigForMid(
+	atsMajorVersion int,
+	profilesCacheKeyConfigParams map[int]map[string]string,
+	server *ServerInfo,
+	dses []RemapConfigDSData,
+	header string,
+) string {
+	midRemaps := map[string]string{}
+	for _, ds := range dses {
+		if ds.Type.IsLive() && !ds.Type.IsNational() {
+			continue // Live local delivery services skip mids
+		}
+
+		if ds.OriginFQDN == nil || *ds.OriginFQDN == "" {
+			log.Warnf("GetServerConfigRemapDotConfigForMid ds '" + ds.Name + "' has no origin fqdn, skipping!") // TODO confirm - Perl uses without checking!
+			continue
+		}
+
+		if midRemaps[*ds.OriginFQDN] != "" {
+			continue // skip remap rules from extra HOST_REGEXP entries
+		}
+
+		// multiple uses of cacheurl and cachekey plugins don't work right in ATS, but Perl has always done it.
+		// So for now, keep track of it, so we can log an error when it happens.
+		hasCacheURL := false
+		hasCacheKey := false
+
+		midRemap := ""
+		if ds.MidHeaderRewrite != nil && *ds.MidHeaderRewrite != "" {
+			midRemap += ` @plugin=header_rewrite.so @pparam=` + MidHeaderRewriteConfigFileName(ds.Name)
+		}
+		if ds.QStringIgnore != nil && *ds.QStringIgnore == tc.QueryStringIgnoreIgnoreInCacheKeyAndPassUp {
+			qstr, addedCacheURL, addedCacheKey := GetQStringIgnoreRemap(atsMajorVersion)
+			if addedCacheURL {
+				hasCacheURL = true
+			}
+			if addedCacheKey {
+				hasCacheKey = true
+			}
+			midRemap += qstr
+		}
+		if ds.CacheURL != nil && *ds.CacheURL != "" {
+			if hasCacheURL {
+				log.Errorln("DeliveryService qstring_ignore and cacheurl both add cacheurl, but ATS cacheurl doesn't work correctly with multiple entries! Adding anyway!")
+			}
+			midRemap += ` @plugin=cacheurl.so @pparam=` + CacheURLConfigFileName(ds.Name)
+		}
+
+		if ds.ProfileID != nil && len(profilesCacheKeyConfigParams[*ds.ProfileID]) > 0 {
+			if hasCacheKey {
+				log.Errorln("DeliveryService qstring_ignore and cachekey params both add cachekey, but ATS cachekey doesn't work correctly with multiple entries! Adding anyway!")
+			}
+			midRemap += ` @plugin=cachekey.so`
+			for name, val := range profilesCacheKeyConfigParams[*ds.ProfileID] {
+				midRemap += ` @pparam=--` + name + "=" + val
+			}
+		}
+		if ds.RangeRequestHandling != nil && *ds.RangeRequestHandling == tc.RangeRequestHandlingCacheRangeRequest {
+			midRemap += ` @plugin=cache_range_requests.so`
+		}
+
+		if midRemap != "" {
+			midRemaps[*ds.OriginFQDN] = midRemap
+		}
+	}
+
+	textLines := []string{}
+	for originFQDN, midRemap := range midRemaps {
+		textLines = append(textLines, "map "+originFQDN+" "+originFQDN+midRemap+"\n")
+	}
+	sort.Strings(textLines)
+
+	text := header
+	for _, line := range textLines {
+		text += line
+	}
```

Comment:
I think the standard way to do this is to use `strings.Join`.

Reply:
Changed.